compiler-rt/lib/tsan/rtl/tsan_platform_linux.cpp

```diff
@@ -292,8 +292,12 @@ static void ReExecIfNeeded(bool ignore_heap) {
     } else {
       Printf(
           "FATAL: ThreadSanitizer: memory layout is incompatible, "
-          "even though ASLR is disabled.\n"
-          "Please file a bug.\n");
+          "even though ASLR is disabled.\n");
+      Printf(
+          "FATAL: This error may occur for programs that use GPU libraries.");
```

Comment:
Why the extra `FATAL`?

Reply:
Users may grep/filter for 'FATAL' in the log, and it would be misleading if they only saw a subset of the lines.

Reply:
OK

Reply:
Why do we need to mention GPU?

I can see this can help some, users, but it also can misleads other users who has issue, but not GPU libraries.
If we can detect The Library, then maybe there is a value in the message.

What is % of total issues do we expect caused by GPU libraries? 
We need a confidence that it's high to blame them.

Reply:
> Why do we need to mention GPU?
> 
> I can see this can help some users, but it also can misleads other users, who has issue, but no GPU libraries.

It won't mislead them if they have no GPU libraries - the next line is "If your program does not use GPU libraries, please file a TSan bug."

> If we can detect The Library, then maybe there is a value in the message.
> 
> What is % of total issues do we expect caused by GPU libraries? We need a confidence that it's high to blame them.

So far, 100% of issues (sample size = 1) :-).

Most non-GPU libraries aren't mapping large host memory buffers before the shadow is allocated.

Reply:
You are changing the lib which maps a lot of memory.

Lets just print generic link to:
https://clang.llvm.org/docs/ThreadSanitizer.html#limitations

and add entry about your discovery